dask/dashboard/vis_app.py

```diff
@@ -0,0 +1,213 @@
+            req_width = g.width * scale
+            req_height = req_width/h2w
+            text_scale = 12/req_width
+        else:
+            scale = self.screen_height/(self.screen_height - 2*self.border)
+            req_width = g.height*h2w * scale
+            req_height = req_width/h2w
+            text_scale = 8/req_height
+
+        left = x_center - req_width/2
+        right = x_center + req_width/2
+        top = y_center + req_height/2
+        bottom = y_center - req_height/2
+        font_size = str(round(text_scale, 2)) + 'em'
+
+        # Plot
+        self.plot = Figure(title=None, x_axis_type=None, y_axis_type=None,
+                           x_range=[left, right], y_range=[bottom, top],
+                           plot_width=self.screen_width,
+                           plot_height=self.screen_height)
+
+        # Font size can't be linked to data... :(
+        func_text = Text(x='x', y='y', text='label', text_align='center',
+                         text_baseline='middle', text_font_size=font_size)
+        data_text = Text(x='x', y='y', text='label', text_align='center',
+                         text_baseline='middle', text_font_size=font_size)
+
+        # Add glyphs
+        self.plot.add_glyph(self.data_source, self.data_glyphs)
+        self.plot.add_glyph(self.data_source, data_text)
+        self.plot.add_glyph(self.func_source, self.func_glyphs)
+        self.plot.add_glyph(self.func_source, func_text)
+        self.plot.add_glyph(self.edge_source, self.edge_glyphs)
+
+        # Update the data sources
+        edge_lk = edge_data(g)
+        data_lk, func_lk = node_data(g, font_size)
+
+        self.data_source.data.update(data_lk)
+        self.func_source.data.update(func_lk)
+        self.edge_source.data.update(edge_lk)
+
+        self.document.clear()
+        self.document.add(self.plot)
+        self.session.store_document(self.document)
+        self.session.store_objects(self.data_source, self.func_source,
+                                   self.edge_source)
+
+    def update_dask(self, dsk, data_lk, func_lk):
+        if self._dsk is not dsk:
+            self.visualize(dsk)
+        self.data_source.data.update(data_lk)
+        self.func_source.data.update(func_lk)
+        self.session.store_objects(self.data_source, self.func_source)
+
+    def scheduler_callback(self, key, dsk, state):
+        sleep(self.pause)
+        self._queue.put(color_nodes(dsk, state))
+
+    def get(self, dsk, result, cache=None, **kwargs):
+        return self._get(dsk, result, cache=cache,
+                         scheduler_callback=self.scheduler_callback)
+
+
+class _Handler(Thread):
+    def __init__(self, vis):
+        super(_Handler, self).__init__()
+        self.queue = vis._queue
+        self.update_dask = vis.update_dask
+        self.daemon = True
+
+    def run(self):
+        while True:
+            task = self.queue.get()
+            self.update_dask(*task)
```

Comment:
I'd recommend to not re-use the glyphs this way... better to just recreate them. I've run into problems reusing glyphs recently but didn't have time to dig into that and formalize/reproduce the issue.